app/src/main/java/org/mozilla/focus/browser/BrowserNavigationOverlay.kt

```diff
@@ -75,13 +92,42 @@ class BrowserNavigationOverlay @JvmOverloads constructor(
                     it.setOnClickListener(this)
                 }
 
+        uiLifecycleCancelJob = Job()
+
+        initTiles()
         setupUrlInput()
         turboButton.isChecked = Settings.getInstance(context).isBlockingEnabled
         navButtonSettings.setImageResource(R.drawable.ic_settings) // Must be set in code for SVG to work correctly.
 
         val tintDrawable: (Drawable?) -> Unit = { it?.setTint(ContextCompat.getColor(context, R.color.tv_white)) }
         navCloseHint.compoundDrawablesRelative.forEach(tintDrawable)
         navUrlInput.compoundDrawablesRelative.forEach(tintDrawable)
+
+        tileContainer.addOnScrollListener(object : RecyclerView.OnScrollListener() {
+            override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
+                val gridLayoutManager = recyclerView?.layoutManager as GridLayoutManager
+                val lastVisibleItem = gridLayoutManager.findLastCompletelyVisibleItemPosition()
+                // We add a scroll offset, revealing the next row to hint that there are more home tiles
+                if (dy > 0 && getFocusedTilePosition() > lastVisibleItem) {
+                    val scrollOffset = TypedValue.applyDimension(
+                            TypedValue.COMPLEX_UNIT_DIP, home_tile.height.toFloat() / 2, context.resources.displayMetrics)
+                    recyclerView.smoothScrollBy(0, scrollOffset.toInt())
+                }
+            }
+        })
+    }
+
+    private fun initTiles() = with (tileContainer) {
+        val homeTiles = mutableListOf<HomeTile>().apply {
+            addAll(BundledTilesManager.getInstance(context).getBundledHomeTilesList())
+            addAll(CustomTilesManager.getInstance(context).getCustomHomeTilesList())
+        }
+
+        adapter = HomeTileAdapter(uiLifecycleCancelJob, homeTiles, loadUrl = { urlStr ->
+            onNonTextInputUrlEntered(urlStr)
```

Comment:
I'm a little confused where onNonTextInputUrlEntered is, because I can't find it in the rest of the file.